pulsar/internal/blocking_queue.go

```diff
@@ -173,17 +206,27 @@ func (bq *blockingQueue) Size() int {
 	return bq.size
 }
 
-func (bq *blockingQueue) Iterator() BlockingQueueIterator {
-	bq.mutex.Lock()
-	defer bq.mutex.Unlock()
-
+func (bq *blockingQueue) iterator() BlockingQueueIterator {
 	return &blockingQueueIterator{
 		bq:      bq,
 		readIdx: bq.headIdx,
 		toRead:  bq.size,
 	}
 }
 
+func (bq *blockingQueue) IterateIfNonEmpty(operation IterateFunc) {
```

Comment:
i think BlockingQueue is thread safe.
but if you run the following code

```go 
q := NewBlockingQueue(3)
q.Put(1)
q.Put(2)
q.Put(3)
iter := q.Iterator()
q.Poll()
q.Put(4)
```
we got 4,1,2 here. but the queue should return 2 before 4.
if some goroutine run concurrently. i think the Iterator is not woking as expected.
so i add an safe IteratorIfNonEmpty to avoid this